js/modules/k6/html/serialize.go

```diff
@@ -0,0 +1,95 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2016 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package html
+
+import (
+	neturl "net/url"
+
+	"github.com/PuerkitoBio/goquery"
+	"github.com/dop251/goja"
+)
+
+type FormValue struct {
+	Name  string
+	Value goja.Value
+}
+
+// nolint: goconst
+func (s Selection) SerializeArray() []FormValue {
+	submittableSelector := "input,select,textarea,keygen"
+	var formElements *goquery.Selection
+	if s.sel.Is("form") {
+		formElements = s.sel.Find(submittableSelector)
+	} else {
+		formElements = s.sel.Filter(submittableSelector)
+	}
+
+	formElements = formElements.FilterFunction(func(i int, sel *goquery.Selection) bool {
+		name := sel.AttrOr("name", "")
+		inputType := sel.AttrOr("type", "")
+		disabled := sel.AttrOr("disabled", "")
+		checked := sel.AttrOr("checked", "")
+
+		return name != "" && // Must have a non-empty name
+			disabled != "disabled" && // Must not be disabled
+			inputType != "submit" && // Must not be a button
+			inputType != "button" &&
+			inputType != "reset" &&
+			inputType != "image" && // Must not be an image or file
+			inputType != "file" &&
+			(checked == "checked" || (inputType != "checkbox" && inputType != "radio")) // Must be checked if it is an checkbox or radio
+	})
+
+	result := make([]FormValue, len(formElements.Nodes))
+	formElements.Each(func(i int, sel *goquery.Selection) {
+		element := Selection{s.rt, sel, s.URL}
+		name, _ := sel.Attr("name")
+		result[i] = FormValue{Name: name, Value: element.Val()}
```

Comment:
A check must be added to look for non-set value here. Right now fields without a `value` attribute gets an `undefined` value. The jQuery docs say:
> Elements that do not contain a value attribute are represented with the empty string value.

So we need to add the equivalent check to what Cheerio is doing here: https://github.com/cheeriojs/cheerio/blob/master/lib/api/forms.js#L47

And please add a test case for this as well.

Reply:
Hm, on a second look this might actually be a bug here https://github.com/loadimpact/k6/blob/master/js/modules/k6/html/html.go#L280. `val()` of an input fields without a `value` attribute returns an empty string in jQuery, whereas that line returns an `goja.Undefined()` at https://github.com/loadimpact/k6/blob/master/js/modules/k6/html/html.go#L264.

Reply:
@robingustafsson nice catch. I already fixed this in #437, but will also backport it to this pull request (because #437 contains way more).

Reply:
Ah great, was gonna look at that PR next :)